File: app/src/main/java/com/cyl/musiclake/ui/music/online/activity/BaiduMusicListActivity.java
package com.cyl.musiclake.ui.music.online.activity;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.cyl.musiclake.R;
import com.cyl.musiclake.base.BaseActivity;
import com.cyl.musiclake.common.Constants;
import com.cyl.musiclake.common.Extras;
import com.cyl.musiclake.data.db.Music;
import com.cyl.musiclake.player.PlayManager;
import com.cyl.musiclake.ui.music.dialog.DownloadDialog;
import com.cyl.musiclake.ui.music.dialog.ShowDetailDialog;
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter;
import com.cyl.musiclake.ui.music.online.contract.BaiduListContract;
import com.cyl.musiclake.ui.music.online.presenter.BaiduListPresenter;
import com.cyl.musiclake.utils.CoverLoader;
import com.cyl.musiclake.utils.FormatUtil;
import com.cyl.musiclake.utils.SizeUtils;
import com.cyl.musiclake.utils.ToastUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * 作者：yonglong on 2016/8/24 10:43
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
@SuppressWarnings("ConstantConditions")
public class BaiduMusicListActivity extends BaseActivity<BaiduListPresenter> implements BaiduListContract.View {

    private static final String TAG = "BaiduMusicListActivity";
    private List<Music> musicList = new ArrayList<>();
    private SongAdapter mAdapter;

    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;

    private View mViewHeader;
    ImageView mIvBackground;
    ImageView mIvCover;
    TextView mTvTitle;
    TextView mTvDate;
    TextView mTvDesc;

    private int mOffset = 0;
    private String title;
    private String type;
    private String desc;
    private String pic;
    private int mCurrentCounter = 0;
    private int TOTAL_COUNTER = 0;
    private int limit = 10;

    @Override
    protected int getLayoutResID() {
        return R.layout.activity_online;
    }

    @Override
    protected void initView() {
        title = getIntent().getStringExtra(Extras.BILLBOARD_TITLE);
        type = getIntent().getStringExtra(Extras.BILLBOARD_TYPE);
        desc = getIntent().getStringExtra(Extras.BILLBOARD_DESC);
        pic = getIntent().getStringExtra(Extras.BILLBOARD_ALBUM);
        setToolbarTitle(title);
        initHeaderView();
    }


    @Override
    protected void initData() {
        mAdapter = new SongAdapter(musicList);
        mAdapter.setEnableLoadMore(true);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(mAdapter);
        mAdapter.bindToRecyclerView(mRecyclerView);
        showHeaderInfo();

        mPresenter.loadOnlineMusicList(type, 10, mOffset);
    }

    @Override
    protected void initInjector() {
        mActivityComponent.inject(this);
    }

    @Override
    protected void listener() {
        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            if (view.getId() != R.id.iv_more) {
                PlayManager.play(position, musicList, Constants.BAIDU + type);
            }
        });
        mAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            Music music = (Music) adapter.getItem(position);
            PopupMenu popupMenu = new PopupMenu(getApplicationContext(), view);
            popupMenu.setOnMenuItemClickListener(item -> {
                switch (item.getItemId()) {
                    case R.id.popup_song_detail:
                        ShowDetailDialog.newInstance(music)
                                .show(getSupportFragmentManager(), getLocalClassName());
                        break;
                    case R.id.popup_add_playlist:
                        ToastUtils.show("暂不支持添加百度音乐");
                        break;
                    case R.id.popup_song_download:
                        DownloadDialog.newInstance(music)
                                .show(getSupportFragmentManager(), getLocalClassName());
                        break;
                }
                return false;
            });
            popupMenu.inflate(R.menu.popup_song_online);
            popupMenu.show();
        });
        mAdapter.setOnLoadMoreListener(() -> mRecyclerView.postDelayed(() -> {
            if (mCurrentCounter < TOTAL_COUNTER) {
                //数据全部加载完毕
                mAdapter.loadMoreEnd();
            } else {
                //成功获取更多数据
                mPresenter.loadOnlineMusicList(type, limit, mOffset);
            }
        }, 1000), mRecyclerView);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return super.onCreateOptionsMenu(menu);
    }


    private void initHeaderView() {
        mViewHeader = LayoutInflater.from(this).inflate(R.layout.activity_online_header, null);
        AbsListView.LayoutParams params = new AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, SizeUtils.dp2px(this, 150));
        mViewHeader.setLayoutParams(params);

        mIvCover = (ImageView) mViewHeader.findViewById(R.id.iv_cover);
        mTvTitle = (TextView) mViewHeader.findViewById(R.id.tv_title);
        mTvDate = (TextView) mViewHeader.findViewById(R.id.tv_update_date);
        mTvDesc = (TextView) mViewHeader.findViewById(R.id.tv_comment);

        mTvDate.setText("最近更新：" + FormatUtil.distime(System.currentTimeMillis()));
    }

    @Override
    public void showLoading() {
    }

    @Override
    public void hideLoading() {
    }

    @Override
    public void showErrorInfo(String msg) {
        ToastUtils.show(this, msg);
        mAdapter.loadMoreFail();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
        }
        return super.onOptionsItemSelected(item);
    }

    public void showHeaderInfo() {
        mTvTitle.setText(title);
//        mTvDate.setText(getString(R.string.recent_update, playlistInfo.getUpdate_date()));
        mTvDesc.setText(desc);
        CoverLoader.loadImageView(this, pic, mIvCover);
        mAdapter.setHeaderView(mViewHeader, 0);
    }

    @Override
    public void showOnlineMusicList(List<Music> musicList) {
        this.musicList.addAll(musicList);
        mAdapter.setNewData(this.musicList);
        mOffset = mOffset + limit;
        mCurrentCounter = mAdapter.getData().size();
        TOTAL_COUNTER = mOffset;
        mAdapter.loadMoreComplete();
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/online/activity/BaiduMusicListActivity.kt
package com.cyl.musiclake.ui.music.online.activity

import android.support.v7.widget.LinearLayoutManager
import android.view.*
import android.widget.AbsListView
import android.widget.ImageView
import android.widget.TextView
import com.cyl.musiclake.R
import com.cyl.musiclake.base.BaseActivity
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.common.Extras
import com.cyl.musiclake.common.NavigationHelper
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.player.PlayManager
import com.cyl.musiclake.ui.music.dialog.BottomDialogFragment
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter
import com.cyl.musiclake.ui.music.online.contract.BaiduListContract
import com.cyl.musiclake.ui.music.online.presenter.BaiduListPresenter
import com.cyl.musiclake.utils.CoverLoader
import com.cyl.musiclake.utils.FormatUtil
import com.cyl.musiclake.utils.ToastUtils
import kotlinx.android.synthetic.main.activity_online.*

/**
 * 作者：yonglong on 2016/8/24 10:43
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
class BaiduMusicListActivity : BaseActivity<BaiduListPresenter>(), BaiduListContract.View {
    private val musicList = mutableListOf<Music>()
    private val mAdapter by lazy { SongAdapter(musicList) }

    private var mViewHeader: View? = null
    private var mIvBackground: ImageView? = null
    private var mIvCover: ImageView? = null
    private var mTvTitle: TextView? = null
    private var mTvDate: TextView? = null
    private var mTvDesc: TextView? = null

    private var mOffset = 0
    private var title: String? = null
    private var type: String? = null
    private var desc: String? = null
    private var pic: String? = null
    private var mCurrentCounter = 0
    private var TOTAL_COUNTER = 0
    private val limit = 10

    override fun getLayoutResID(): Int {
        return R.layout.activity_online
    }

    override fun initView() {
        title = intent.getStringExtra(Extras.BILLBOARD_TITLE)
        type = intent.getStringExtra(Extras.BILLBOARD_TYPE)
        desc = intent.getStringExtra(Extras.BILLBOARD_DESC)
        pic = intent.getStringExtra(Extras.BILLBOARD_ALBUM)
        initHeaderView()
    }

    override fun setToolbarTitle(): String? {
        title = intent.getStringExtra(Extras.BILLBOARD_TITLE)
        return title
    }

    override fun initData() {
        mAdapter.setEnableLoadMore(true)
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = mAdapter
        mAdapter.bindToRecyclerView(recyclerView)
        showHeaderInfo()

        type?.let { mPresenter?.loadOnlineMusicList(it, limit, mOffset) }
    }

    override fun initInjector() {
        mActivityComponent.inject(this)
    }

    override fun listener() {
        mAdapter.setOnItemClickListener { _, view, position ->
            if (view.id != R.id.iv_more) {
                PlayManager.play(position, musicList, Constants.BAIDU + type)
                mAdapter.notifyDataSetChanged()
                NavigationHelper.navigateToPlaying(this)
            }
        }
        mAdapter.setOnItemChildClickListener { adapter, view, position ->
            val music = adapter.getItem(position) as Music?
            BottomDialogFragment.newInstance(music, Constants.OP_ONLINE).show(this)
        }
        mAdapter.setOnLoadMoreListener({
            recyclerView.postDelayed({
                if (mCurrentCounter < TOTAL_COUNTER) {
                    //数据全部加载完毕
                    mAdapter.loadMoreEnd()
                } else {
                    //成功获取更多数据
                    type?.let { mPresenter?.loadOnlineMusicList(it, limit, mOffset) }
                }
            }, 1000)
        }, recyclerView)
    }

    private fun initHeaderView() {
        mViewHeader = LayoutInflater.from(this).inflate(R.layout.activity_online_header, null)
        val params = AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        mViewHeader?.layoutParams = params

        mIvCover = mViewHeader?.findViewById<View>(R.id.iv_cover) as ImageView
        mTvTitle = mViewHeader?.findViewById<View>(R.id.tv_title) as TextView
        mTvDate = mViewHeader?.findViewById<View>(R.id.tv_update_date) as TextView
        mTvDesc = mViewHeader?.findViewById<View>(R.id.tv_comment) as TextView

        mTvDate?.text = getString(R.string.recently_update, FormatUtil.distime(System.currentTimeMillis()))
    }

    override fun showLoading() {
        super.showLoading()
    }

    override fun hideLoading() {
        super.hideLoading()
    }

    override fun showErrorInfo(msg: String) {
        ToastUtils.show(this, msg)
        mAdapter.loadMoreFail()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showHeaderInfo() {
        mTvTitle?.text = title
        //        mTvDate.setText(getString(R.string.recent_update, playlistInfo.getUpdate_date()));
        mTvDesc?.text = desc
        CoverLoader.loadImageView(this, pic, mIvCover)
        mAdapter.setHeaderView(mViewHeader, 0)
    }

    override fun showOnlineMusicList(musicList: List<Music>) {
        this.musicList.addAll(musicList)
        mAdapter.setNewData(this.musicList)
        mOffset += limit
        mCurrentCounter = mAdapter.data.size
        TOTAL_COUNTER = mOffset
        mAdapter.loadMoreComplete()
    }

    companion object {
        private val TAG = "BaiduMusicListActivity"
    }
}
